core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,7 +64,23 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // Evaluate partitions in parallel. Partitions of each rdd will be cached by the `partitions`
+  // val in `RDD`.
+  private[spark] lazy val parallelPartitionEval: Boolean = {
```

Comment:
You're right that there are several instances of InputFormat and those aren't shared across threads. The problem is that Hadoop uses reflection to get InputFormat instances, which means the way for an InputFormat to cache intermediate results is to use a static cache available across instances. I don't think this is a huge risk, but I think it warrants being able to disable UnionRDD concurrency.